File: safe_vec/safe_vec.cpp
#include <array>
#include <iostream>
#include <new>
#include <vector>

auto constexpr SIZE = 16384;

auto loop() -> std::pair<size_t, std::bad_alloc> {
    auto v = std::vector<int>();
    try {
        for (;;) {
            v.push_back(0);
        }
    } catch (std::bad_alloc e) {
        return {v.size(), e};
    }
    throw "unreachable";
}

auto main(int argc, char** argv) -> int {
    auto [len, e] = loop();
    std::cout << " [cpp] we got to " << len
              << " before the os kicked us out because of " << e.what()
              << std::endl;
}

File: safe_vec/safe_vec.rs
use std::collections::TryReserveError;

trait VecExt<V> {
    fn push_checked(&mut self, v: V) -> Result<(), TryReserveError>;
}

impl<V> VecExt<V> for Vec<V> {
    fn push_checked(&mut self, v: V) -> Result<(), TryReserveError> {
        self.try_reserve(1)?;
        Ok(self.push(v))
    }
}

fn main() {
    let mut v = Vec::new();
    let e = loop {
        if let Err(e) = v.push_checked(0) {
            break e;
        }
    };
    let len = v.len();
    drop(v);
    println!("[rust] we got to {len} before the os kicked us out because {e}");
}
